### PR TITLE
drivers: wifi: esp_at: use dormant states

### PR DESCRIPTION
**Pull request by JordanYates**
Update ESP AT modem driver to control the network interface dormant state based on the network connection status, as described in the docs.

As per [here](https://docs.zephyrproject.org/latest/connectivity/networking/api/net_if.html#network-interface-state-management) and: ebdf7da914e887e64ef7f166c475e12f8fe5ec3a
Related to #56550

### CODE REVIEW

**Comment by JordanYates** [comment]
I would appreciate input from @rlubos or any other Nordic contributors, as my end goal is for this and the [nRF7002 driver](https://github.com/nrfconnect/sdk-nrf/tree/main/drivers/wifi/nrf700x/zephyr/src) to generate the same events/states as closely as possible.

**Comment by rlubos** [comment]
For `nrf7002` driver we've moved in the same direction as this PR - we set carrier to ON when Wi-Fi driver is ready (initialized) and then use dormant state to reflect network association status (see https://github.com/nrfconnect/sdk-nrf/pull/11400).

**Comment by JordanYates** [comment]
> For `nrf7002` driver we've moved in the same direction as this PR - we set carrier to ON when Wi-Fi driver is ready (initialized) and then use dormant state to reflect network association status (see [nrfconnect/sdk-nrf#11400](https://github.com/nrfconnect/sdk-nrf/pull/11400)).

That is great news, I was looking at an old version of the driver

File: drivers/wifi/esp_at/esp.c [hunk 1]
```diff
@@ -975,7 +980,8 @@ static int esp_mgmt_connect(const struct device *dev,
 	struct esp_data *data = dev->data;
 	int len;
 
-	if (!net_if_is_up(data->net_iface)) {
+	if (!net_if_is_carrier_ok(data->net_iface) ||
```
Comment (mniestroj) [inline drivers/wifi/esp_at/esp.c:983]: Please also update `esp_mgmt_scan()` to check for `net_if_is_carrier_ok()`, so that scan works in dormant mode.

**Review by mniestroj — COMMENTED** [review]
This change looks reasonable. However, during testing I noticed that following flow (that I used so far) no longer works:
- wait for `NET_EVENT_IF_UP`, to make sure ESP-AT driver is initialized and ready for WiFi mgmt commands
- issue WiFi connect

After applying this PR, there is no way to get notified about ESP-AT WiFi interface being ready to consume WiFi mgmt commands (like connect or scan). I guess we would need something like `NET_EVENT_IF_CARRIER_OK` (to be introduced) event to make that work.

**Comment by JordanYates** [comment]
> This change looks reasonable. However, during testing I noticed that following flow (that I used so far) no longer works:

That is by design (it is how I used to work as well). `NET_EVENT_IF_UP` should only be triggered when the interface is ready for sending data.

In combination with #59331, you will know the ESP-AT driver is initialized and ready for WiFi mgmt commands when the device init function returns and `device_is_ready` returns true.

File: drivers/wifi/esp_at/esp.c (hunk 1 above)
Comment (JordanYates) [reply drivers/wifi/esp_at/esp.c:983]: Done

**Comment by carlescufi** [comment]
@mniestroj could you please take another look? Thanks!

**Review by mniestroj — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Comment by WilliamGFish** [comment]
So how do we initialise the ESP32 AT modem now??? As per the comments above it fails to start...
Billy..

**Comment by JordanYates** [comment]
The AT modem will be initialised with `carrier=on`, `dormant=true` (assuming the device init function succeeds).
Once `NET_REQUEST_WIFI_CONNECT` is sent to the modem, and it successfully connects to an AP, the dormant state will be automatically turned off. At that point, `NET_EVENT_IF_UP ` will be raised.

You should NOT be manually controlling either the carrier state or dormant states at the application layer. They refer to physical states of the device, not the desire of the application.

**Pull request merged**